lib/parsers.js

```diff
@@ -142,9 +150,16 @@ exports.getContextFromRequest = function (req, conf, type) {
 
   if (haveBody) {
     if (captureBody) {
+      body = removeKeysFromPostedFormVariables(body, req.headers, conf.sanitizeFieldNamesRegExp)
+
+      if (Buffer.isBuffer(body)) {
```

Comment:
This conditional changes how the agent handles the raw/Buffer body parsing.  Previously, if a customer user the raw/buffer middleware, we'd send up data that looked like this

    '{"type":"Buffer","data":[49,50,51,52,53,54,55]}'

i.e. we weren't really handling this middleware in a way that was useful to users.  This conditional turns the payload Buffer into its string representation.

Reply:
I'm uncomfortable making the assumption that the POST'd body *is* meaningfully representable as a string. E.g. here is me with a simple file upload endpoint (using express):

```
var rawParser = bodyParser.raw()
app.post('/see-my-body', rawParser, (req, res) => {
  console.log('XXX req.headers', req.headers);
  console.log('XXX req.body', req.body);
  res.end('I see you')
}
```

and POSTing a small ~60kB [kitty image](https://www.litter-robot.com/blog/wp-content/uploads/2016/08/shutterstock_191770991-scared-cat-900x600.jpg) to it:

```
% curl   -Sis http://localhost:3000/see-my-body --data-binary @kitty.jpg -H content-type:application/octet-stream
HTTP/1.1 100 Continue

HTTP/1.1 200 OK
X-Powered-By: Express
Date: Fri, 18 Dec 2020 20:14:15 GMT
Connection: keep-alive
Content-Length: 9

I see you
```

The log from my app's endpoint is:

```
XXX req.headers { host: 'localhost:3000',
  'user-agent': 'curl/7.64.1',
  accept: '*/*',
  'content-type': 'application/octet-stream',
  'content-length': '69930',
  expect: '100-continue' }
XXX req.body <Buffer ff d8 ff e0 00 10 4a 46 49 46 00 01 01 01 01 2c 01 2c 00 00 ff db 00 43 00 06 04 04 05 04 04 06 05 05 05 06 06 06 07 09 0e 09 09 08 08 09 12 0d 0d 0a ... >
```

*with* this `body = body.toString()` the data sent to APM server looks like this:

![Screen Shot 2020-12-18 at 12 13 47 PM](https://user-images.githubusercontent.com/46866/102658125-cfc31580-412b-11eb-9477-0f57a0f69aec.png)

What about considering the capturing of a *binary* body (leading to the current `'{"type":"Buffer","data":[49,50,51,52,53,54,55]}'`) to be a bug? Instead we change the definition of `captureBody: true` to either (a) exclude a binary (i.e. Buffer.isBuffer) body, or explicitly encode it as something like `<Buffer>`:

```js
if (haveBody) {
    if (!captureBody) {
        // Aside: I'm not sure why the agent calls this "REDACTED" here. If
        // the config is `captureBody: false`, then I would just leave
        // context.body = undefined.
        context.body = '[REDACTED]'
    } else if (Buffer.isBuffer(body)) {
        context.body = '<Buffer>'
    } else {
        ...
    }
}
```

Reply:
Agreed -- but I'll tackle this in a separate PR: https://github.com/elastic/apm-agent-nodejs/issues/1925